asmcomp/cmm_helpers.ml

```diff
@@ -1390,10 +1476,27 @@ let make_switch arg cases actions dbg =
     if length >= 2
     then begin
       match const_actions.(cases.(0)), const_actions.(cases.(1)) with
-      | Cint v0, Cint v1 ->
-          let slope = Nativeint.sub v1 v0 in
+      | Cint n0, Cint n1 ->
+          (* The right-hand-sides are tagged, so we can translate them
+             back to OCaml integers without loss of information, to
+             compute the offset and slope on OCaml integers.
+
+             For example, consider the identity function on OCaml integers
+               0 -> 0
+               1 -> 1
+               2 -> 2
+             If we computed the slope with native integers on the
+             right-hand-side, we would see
+               0 -> 1n
+               1 -> 3n
+               2 -> 5n
+             and compute offset=1n, slope=2n.
+             We want offset=0, slope=1 instead.
+          *)
+          let v0, v1 = untag_const n0, untag_const n1 in
+          let slope = v1 - v0 in
           let check i = function
-            | Cint v -> v = Nativeint.(add (mul (of_int i) slope) v0)
+            | Cint v -> untag_const v = (slope * i + v0)
```

Comment:
Nitpick, but I'd prefer if you kept consistent variable names: `n` is a Cmm integer, `v` is its untagged version.

Reply:
done, thanks!